wordpress-mobile/WordPress-iOS
# Adds gif support to reader detail content and reader comments

This is an almost-complete PR that adds gif support to the reader detail content, reader detail full screen preview (`WPImageViewController`), and the reader comments:

![gif_reader2](https://user-images.githubusercontent.com/154014/40214551-0bb4897e-5a21-11e8-86e3-27fc0872fecf.gif)

It's basically ready for you to start taking a look at @etoledom.

Instead of using `WPTableImageSource` to load the images, `WPRichTextImage` now uses our new `ImageLoader` for all images (animated or not).  `ImageLoader` was updated to to now allow a photon quality var to be set — which is `65` for within the reader detail (and defaults to `80` everywhere else like previously).

Trying to reduce memory is the name of the game here. I put in a few safeguards like:

* Selecting the `small` gif playback strategy
* `AnimatedImageCache` is now listening for memory warnings and dumping the cache if it occurs
* As mentioned ☝️, `ImageLoader` sets photon quality to 65
* Clearing data from `CachedAnimatedImageView` as soon as possible

Even with all that ☝️, memory usage **is** higher than our previous static, non-animated gif display (assuming you view a lot of reader posts containing gifs). The question is if this implementation is acceptable in terms of memory usage.

Fixes #9253 
Fixes #9331  

## WIP

1. Like the previous reader image loading implementation, there isn't an loading indicator for images. The image size is set AFTER the image is loaded. I am not sure if this is best approach now, and will most likely add the indicator back in. 😄 

2. There is a image sizing issue while scrolling through reader comments. I'll patch it asap.

## To Test:

**Note 1:** Try to test on device since that will give us an accurate gauge of mem usage.

**Note 2:** Be sure to also test a slow network, not just wifi.

**Note 3:** iPads!!!  (Ok, ok no more notes!)

1. Locate reader posts with gifs, heck, locate reader posts with **ALL** kinds of media. Make sure it all works. Hawk the memory usage while you are doing this.

2. Next, find some reader comments that contain gifs (ask me if you need help finding some 😄 ). Verify they display correctly.
